stacksCollectible/index.ts

```diff
@@ -0,0 +1,145 @@
+    const nftArray: Array<NonFungibleToken> = [];
+
+    const batch = nfts.slice(index, index + batchSize);
+    batch.forEach((nft) => {
+      const principal: string[] = nft.asset_identifier.split('::');
+      const contractInfo: string[] = principal[0].split('.');
+      const contractId = principal[0];
+
+      if (contractInfo[1] === 'bns') {
+        //no further data required for BNS, arrange into collection
+        const bnsCollection: StacksCollectionData = {
+          collection_id: 'bns',
+          collection_name: 'BNS Names',
+          total_nft: 1,
+          thumbnail_nfts: [nft],
+          all_nfts: [nft],
+        };
+        collectionRecord[contractInfo[1]] = bnsCollection;
+      } else {
+        nftDetailPromises.push(getNftDetail(nft.value.repr.replace('u', ''), contractInfo[0], contractInfo[1]));
+        collectionDataPromises.push(getNftsCollectionData(contractId));
+        nftArray.push(nft);
+      }
+    });
+    const resolvedNftDetailPromises = await Promise.all(nftDetailPromises);
+    const resolvedCollectionDataPromises = await Promise.all(collectionDataPromises);
+
+    for (let i = 0; i < nftArray.length; i++) {
+      const nft = nftArray[i];
+      const principal: string[] = nft.asset_identifier.split('::');
+      const contractInfo: string[] = principal[0].split('.');
+      const contractId = principal[0];
+
+      if (contractInfo[1] === 'bns') continue; //collection already assigned
+
+      const nftDetail = resolvedNftDetailPromises[i];
+      const collectionData = resolvedCollectionDataPromises[i];
+
+      if (nftDetail) nft.data = nftDetail.data;
+
+      //group NFTs into collections
+      if (collectionRecord[contractId]) {
+        const data = collectionRecord[contractId];
+
+        data.all_nfts.push(nft);
+        if (data.total_nft < 4) {
+          data?.thumbnail_nfts.push(nft);
+        }
+        data.total_nft += 1;
+      } else {
+        collectionRecord[contractId] = {
+          collection_id: contractId,
+          collection_name: collectionData?.collection?.name ?? contractId,
+          total_nft: 1,
+          all_nfts: [nft],
+          thumbnail_nfts: [nft],
+          floor_price: collectionData?.collection?.floorItem?.price
+            ? microstacksToStx(new BigNumber(collectionData?.collection?.floorItem?.price)).toNumber()
+            : 0,
+        };
+      }
+    }
+  }
+
+  const nftCollectionList = Object.values(collectionRecord);
+
+  //sort according to total nft in a collection
+  nftCollectionList.sort((a, b) => {
+    //place bns collection at the bottom of nft list
+    if (a.collection_id === 'bns' || b.collection_id === 'bns') return -1;
+    return b.total_nft - a.total_nft;
+  });
+
+  const requiredSortedCollectionsData = nftCollectionList?.slice(offset, offset + limit);
+
+  return {
+    offset,
+    limit,
+    total: nftCollectionList.length,
+    results: requiredSortedCollectionsData,
+  };
```

Comment:
so every time the user requests the next page, this function will refetch the entire dataset from the gamme API, sort it, and then return the requested page.

should the responses be cached between page requests?

Reply:
need @victorkirov's input here. because the extension/app will cache the returned pages, but not each request that core makes to gamma API. 

so current implementation would be something like:
1. user loads NFTs tab:
  -> core function paginates through all contracts
  -> core function calls collection endpoint for each contract
  -> core function sorts and returns page 1
  -> extension caches page 1
2. user clicks load more on NFTs tab:
  -> core function paginates through all contracts (again)
  -> core function calls collection endpoint for each contract (again)
  -> core function sorts and returns page 2
  -> extension caches page 2

Reply:
Yeah, this will definitely need caching within core. A few notes on this file:
- the call to `getAllNftContracts` should be cached for a reasonable TTL, even if it's just a plain in-memory cache. We don't need to get the full list each time.
- `getNftCollections` should be broken down into smaller functions. There are clear parts it can be broken down into to make the whole function easier to digest.
- `nftCollectionList` should be cached at the end and, when this function is called again, we just take a page out of the cached value to form the `requiredSortedCollectionsData` until the TTL expires.

Reply:
We might not even need to cache `getAllNftContracts` if we're caching `nftCollectionList`

Reply:
what's a reasonable TTL for storing this data? 2 hours or should it be something like a day

Reply:
It will be an in-memory cache, so it doesn't really matter too much. Maybe 1 hour? The moment they close the extension/mobile app, the cache will be lost. The important bit for the cache is to not keep on hitting the API when they try to load more NFTs.

Reply:
I think 1 hour might be too long for mobile. consider the following use cases:

1. user is navigating between tabs, and paginating stx NFTs
2. user has just bought a new NFT, and wants to see it

1 hour TTL will cover 1. but not be great for 2. on mobile. user would have to close the app and restart it to force a refetch

5mins TTL should still cover 1. but also 2.